pkg/apis/discovery/types.go

```diff
@@ -107,6 +107,12 @@ type Endpoint struct {
 	//   endpoint is located. This should match the corresponding node label.
 	// +optional
 	Topology map[string]string
+	// TerminationGracePeriodSeconds is the termination grace period (in seconds) the
+	// endpoint is given to gracefully shutdown. Consumers should use this field to determine
+	// how long existing connections to an endpoint should be kept open after the endpoint
+	// is deleted.
+	// +optional
+	TerminationGracePeriodSeconds *int64
```

Comment:
Something else I considered:
```
DrainUntil metav1.Duration
```

But that would mean we depend on knowing the pod deletion timestamp from the endpoint slice controller which may not always be the case (for example: if pods get deleted during controller manager restart).

Reply:
I wonder if we need duration information: I think a Terminating/Draining state would be enough:
- pod is created and becomes ready
- triggers IPVS realserver addition
- pod enters terminating state
- the Terminating/Draining state becomes True
- triggers IPVS realserver weight=0
- the pod is removed from the slice
- triggers IPVS realserver removal

That would mean having the pod in the slice until it's deleted though
What do you think?

Reply:
OK after reading the full PR, I get it
Doing this way allows to not modify the overall behavior of the controller by keeping endpoints in the slice when they are Terminating
It's simple, I like it

The approach I suggested is slightly better because if a pod is deleted before graceful termination ends (which should be the case for most applications gracefully shutting down) we'd be able to immediately remove the real server.

But since it's a change in the api contract it will be harder to put in place and maintain compatibility

Reply:
The thing is we already have a way to have different subsets in endpoints: `addresses` and `notReadyAddresses` but we don't use it and simply iterate over both: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/endpoints.go#L340-L378
We could always publish NotReady endpoints and add the logic to proxier implementations but we may break other controllers relying on this (kubedns/coredns?)

Reply:
What would be great would be to slightly change the logic here: https://github.com/kubernetes/kubernetes/blob/716b62a889f3ec9b3c78ea0fa62c260a7021c1c1/pkg/controller/endpointslice/utils.go#L74-L79

We'd publish each endpoint with its condition (Ready/Notready) and would leave the decision of using NotReady pods to the consumer: any consumer of endpoints would (probably) also watch services and could adapt its behavior based on `publishNotReadyAddresses`:

- new endpoint, NotReady: add to realservers based on publishNotReadyAddresses
- existing endpoint, Ready: add to realservers with weight 1
- existing endpoint becoming not ready: weight set to 0 unless publishNotReadyAddresses
- deleted endpoint: remove from realservers

Reply:
@robscott given EndpointSlice is still in Beta for controllers and Alpha for kube-proxy do you think the idea I described above would be possible?

Reply:
@lbernail I think this seems reasonable from the perspective of EndpointSlice, especially with how little Andrew's approach changes. I'd initially envisioned an approach that just had a `Terminating` bool, but as you've already pointed out, that requires more significant changes to the controller logic and would result in endpoints sticking around longer within an EndpointSlice. I can also see the merits of the larger change, but think we'd need to run it by a broader audience to ensure the potential benefits are worthwhile.

Reply:
> that requires more significant changes to the controller logic and would result in endpoints sticking around longer within an EndpointSlice.

After going through the controller code and running some tests it actually seems like `EndpointSlice` keeps an endpoint of a Pod even if it's `Terminating` so I think adding a `Terminating` bool as part of `EndpointCondition` would be a small change. Is that behavior expected or should it be like `Endpoints` where terminating pods should not be included endpoints list?

Reply:
This thread is split now between the KEP and the PR.

Can we move it to the KEP?

Reply:
BTW, it seems that if EPSlice retains pods that are terminating that's a bug?  That's not compatible with Endpoints.

@robscott

Reply:
> BTW, it seems that if EPSlice retains pods that are terminating that's a bug? That's not compatible with Endpoints.

Thanks for clarifying, I had a local branch to fix this, opened a PR here https://github.com/kubernetes/kubernetes/pull/89056